functions/src/services/recommendation/recommenders/betaBlockersRecommender.ts

```diff
@@ -0,0 +1,88 @@
+//
+// This source file is part of the ENGAGE-HF project based on the Stanford Spezi Template Application project
+//
+// SPDX-FileCopyrightText: 2023 Stanford University
+//
+// SPDX-License-Identifier: MIT
+//
+
+import { Recommender } from './recommender.js'
+import {
+  MedicationRecommendationCategory,
+  type MedicationRecommendation,
+} from '../../../models/medicationRecommendation.js'
+import { MedicationClassReference, MedicationReference } from '../../codes.js'
+import { ContraindicationCategory } from '../../contraindication/contraindicationService.js'
+import { type RecommendationInput } from '../recommendationService.js'
+
+export class BetaBlockerRecommender extends Recommender {
+  // Methods
+
+  compute(input: RecommendationInput): MedicationRecommendation[] {
+    const currentMedication = this.findCurrentMedication(input.requests, [
+      MedicationClassReference.betaBlockers,
+    ])
+    if (!currentMedication) return this.computeNew(input)
+
+    if (this.isTargetDoseReached(currentMedication))
+      return this.createRecommendation(
+        currentMedication,
+        undefined,
+        MedicationRecommendationCategory.targetDoseReached,
+      )
+
+    return this.createRecommendation(
+      currentMedication,
+      undefined,
+      MedicationRecommendationCategory.personalTargetDoseReached,
+    )
+  }
+
+  // Helpers
+
+  private computeNew(input: RecommendationInput): MedicationRecommendation[] {
+    const contraindicationCategory =
+      this.contraindicationService.checkMedicationClass(
+        input.contraindications,
+        MedicationClassReference.betaBlockers,
+      )
+
+    switch (contraindicationCategory) {
+      case ContraindicationCategory.severeAllergyIntolerance:
+      case ContraindicationCategory.allergyIntolerance:
+        return []
+      case ContraindicationCategory.clinicianListed:
+        return this.createRecommendation(
+          undefined,
+          MedicationReference.carvedilol,
+          MedicationRecommendationCategory.noActionRequired,
+        )
+      case ContraindicationCategory.none:
+        break
+    }
+
+    const medianSystolic = input.vitals.systolicBloodPressure.at(
+      input.vitals.systolicBloodPressure.length / 2,
+    )?.value
+    const medianHeartRate = input.vitals.heartRate.at(
+      input.vitals.heartRate.length / 2,
+    )?.value
```

Comment:
Not sure if I understand this code in a right way, but does this actually calculate the median of the values? I just seems to access the middle element?

Seems like this is also applying to other parts of the code.

Reply:
used the extracted version and that one should work fine.